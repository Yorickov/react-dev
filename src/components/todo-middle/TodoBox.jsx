import React, { Component } from 'react';
import update from 'immutability-helper';
import Item from './Item.jsx';
import routes from '../../lib/routes';
import api from '../../lib/api';

class TodoBox extends Component {
  state = { newTaskText: '', tasks: [] };

  handleChangeText = ({ target: { value } }) => {
    this.setState({ newTaskText: value });
  };

  handleSubmitForm = async (e) => {
    e.preventDefault();
    const { newTaskText, tasks } = this.state;
    const response = await api.post(routes.tasksPath(), { body: newTaskText });
    const newTasks = [response.data, ...tasks];
    this.setState({ newTaskText: '', tasks: newTasks });
  };

  handleFinishTask = (id) => async (e) => {
    e.preventDefault();
    await api.patch(routes.finishTaskPath(id));

    const { tasks } = this.state;
    const index = tasks.findIndex((t) => t.id === id);
    const updatedTasks = update(tasks, { [index]: { $merge: { state: 'finished' } } });
    this.setState({ tasks: updatedTasks });
  };

  handleActivateTask = (id) => async (e) => {
    e.preventDefault();
    await api.patch(routes.activateTaskPath(id));

    const { tasks } = this.state;
    const index = tasks.findIndex((t) => t.id === id);
    const updatedTasks = update(tasks, { [index]: { $merge: { state: 'active' } } });
    this.setState({ tasks: updatedTasks });
  };

  renderActiveTasks(activeTasks) {
    return (
      <div className="todo-active-tasks">
        {activeTasks
          .map((t) => <Item key={t.id} task={t} onChangeTask={this.handleFinishTask(t.id)}/>)}
      </div>
    );
  }

  renderFinishedTasks(finishedTasks) {
    return (
      <div className="todo-finished-tasks">
        {finishedTasks
          .map((t) => <Item key={t.id} task={t} onChangeTask={this.handleActivateTask(t.id)}/>)}
      </div>
    );
  }

  resetTasks = async () => {
    const response = await api.get(routes.tasksPath());
    this.setState({ tasks: response.data });
  }

  componentDidMount() {
    this.resetTasks();
  }

  renderForm() {
    const { newTaskText } = this.state;

    return (
      <form className="todo-form" onSubmit={this.handleSubmitForm}>
        <div className="mb-3">
          <input
            type="text"
            onChange={this.handleChangeText}
            value={newTaskText}
            required
            className="form-control mr-3"
            placeholder="I am going..."
          />
        </div>
        <button type="submit" className="btn btn-primary">add</button>
      </form>
    );
  }

  render() {
    const { tasks } = this.state;
    const activeTasks = tasks.filter((t) => t.state === 'active');
    const finishedTasks = tasks.filter((t) => t.state === 'finished');

    return (
      <div>
        <div className="mb-3">
          {this.renderForm()}
        </div>
        {activeTasks.length > 0 && this.renderActiveTasks(activeTasks)}
        {finishedTasks.length > 0 && this.renderFinishedTasks(finishedTasks)}
      </div>
    );
  }
}

export default TodoBox;
