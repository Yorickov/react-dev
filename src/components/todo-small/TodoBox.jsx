import { uniqueId } from 'lodash';
import React, { useState } from 'react';
import Item from './Item.jsx';

const TodoBox = () => {
  const [newTaskText, setText] = useState('');
  const [tasks, setTasks] = useState([]);

  const handleChangeTask = ({ target: { value } }) => {
    setText(value);
  };

  const handleSubmitForm = (e) => {
    e.preventDefault();
    const newTask = { id: uniqueId(), text: newTaskText };
    setTasks([newTask, ...tasks]);
    setText('');
  };

  const handleRemoveTask = (removingId) => (e) => {
    e.preventDefault();
    setTasks(tasks.filter(({ id }) => id !== removingId));
  };

  const renderForm = () => (
    <form onSubmit={handleSubmitForm} className="todo-form row">
      <div className="col-11">
        <input
          type="text"
          onChange={handleChangeTask}
          value={newTaskText}
          required
          className="form-control mr-3"
          placeholder="I am going..."
        />
      </div>
      <button type="submit" className="btn btn-primary col-1">add</button>
    </form>
  );

  return (
    <div>
      <div className="mb-3">
        {renderForm()}
      </div>
      {tasks.map((task) => (
        <div key={task.id}>
          <Item task={task} onRemove={handleRemoveTask(task.id)} />
          <hr />
        </div>
      ))}
    </div>
  );
};

export default TodoBox;
