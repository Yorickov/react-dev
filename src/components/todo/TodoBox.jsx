import { uniqueId } from 'lodash';
import React, { Component } from 'react';
import Item from './Item.jsx';

class TodoBox extends Component {
  state = { newTaskText: '', tasks: [] };

  handleChangeTask = ({ target: { value } }) => {
    this.setState({ newTaskText: value });
  };

  handleRemoveTask = (removingId) => (e) => {
    e.preventDefault();
    const { tasks } = this.state;
    this.setState({ tasks: tasks.filter(({ id }) => id !== removingId) });
  };

  handleSubmitForm = (e) => {
    e.preventDefault();
    const { tasks, newTaskText } = this.state;
    const newTask = { id: uniqueId(), text: newTaskText };
    this.setState({ newTaskText: '', tasks: [newTask, ...tasks] });
  };

  renderForm() {
    const { newTaskText } = this.state;

    return (
      <form action="" onSubmit={this.handleSubmitForm} className="todo-form row">
        <div className="col-11">
          <input
            type="text"
            onChange={this.handleChangeTask}
            value={newTaskText}
            required
            className="form-control mr-3"
            placeholder="I am going..."
          />
        </div>
        <button type="submit" className="btn btn-primary col-1">add</button>
      </form>
    );
  }

  render() {
    const { tasks } = this.state;

    return (
      <div>
        <div className="mb-3">
          {this.renderForm()}
        </div>
        {tasks.map((task) => (
          <div key={task.id}>
            <Item task={task} onRemove={this.handleRemoveTask(task.id)} />
            <hr />
          </div>
        ))}
      </div>
    );
  }
}

export default TodoBox;
