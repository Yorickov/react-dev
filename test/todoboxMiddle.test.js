import React from 'react';
import timeout from 'timeout-then';
import TodoBox from '../src/components/todo-middle/TodoBox.jsx';

it('TodoBox', async () => {
  const mockEvent = { preventDefault: jest.fn() };

  const wrapper = mount(<TodoBox />);
  expect(wrapper.render()).toMatchSnapshot();

  const input = wrapper.find('input');
  const form = wrapper.find('form');

  input.simulate('change', { target: { value: 'first task' } });
  expect(wrapper.render()).toMatchSnapshot();

  form.simulate('submit', mockEvent);
  await timeout(400);
  expect(wrapper.render()).toMatchSnapshot();

  input.simulate('change', { target: { value: 'second task' } });
  expect(wrapper.render()).toMatchSnapshot();

  form.simulate('submit', mockEvent);
  await timeout(400);
  expect(wrapper.render()).toMatchSnapshot();

  const activeTask = wrapper.find('.todo-active-tasks .todo-task').at(0);
  activeTask.simulate('click');
  await timeout(400);
  wrapper.update();
  expect(wrapper.render()).toMatchSnapshot();

  const finishedTask = wrapper.find('.todo-finished-tasks .todo-task');
  finishedTask.simulate('click');
  await timeout(400);
  wrapper.update();
  expect(wrapper.render()).toMatchSnapshot();
});
