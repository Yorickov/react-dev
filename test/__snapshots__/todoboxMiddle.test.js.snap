// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoBox 1`] = `
<div>
  <div
    class="mb-3"
  >
    <form
      class="todo-form"
    >
      <div
        class="mb-3"
      >
        <input
          class="form-control mr-3"
          placeholder="I am going..."
          required=""
          type="text"
          value=""
        />
      </div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        add
      </button>
    </form>
  </div>
</div>
`;

exports[`TodoBox 2`] = `
<div>
  <div
    class="mb-3"
  >
    <form
      class="todo-form"
    >
      <div
        class="mb-3"
      >
        <input
          class="form-control mr-3"
          placeholder="I am going..."
          required=""
          type="text"
          value="first task"
        />
      </div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        add
      </button>
    </form>
  </div>
</div>
`;

exports[`TodoBox 3`] = `
<div>
  <div
    class="mb-3"
  >
    <form
      class="todo-form"
    >
      <div
        class="mb-3"
      >
        <input
          class="form-control mr-3"
          placeholder="I am going..."
          required=""
          type="text"
          value=""
        />
      </div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        add
      </button>
    </form>
  </div>
  <div
    class="todo-active-tasks"
  >
    <div
      class="row"
    >
      <div
        class="col-1"
      >
        1
      </div>
      <div
        class="col"
      >
        <a
          class="todo-task"
          href="#"
        >
          first task
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`TodoBox 4`] = `
<div>
  <div
    class="mb-3"
  >
    <form
      class="todo-form"
    >
      <div
        class="mb-3"
      >
        <input
          class="form-control mr-3"
          placeholder="I am going..."
          required=""
          type="text"
          value="second task"
        />
      </div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        add
      </button>
    </form>
  </div>
  <div
    class="todo-active-tasks"
  >
    <div
      class="row"
    >
      <div
        class="col-1"
      >
        1
      </div>
      <div
        class="col"
      >
        <a
          class="todo-task"
          href="#"
        >
          first task
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`TodoBox 5`] = `
<div>
  <div
    class="mb-3"
  >
    <form
      class="todo-form"
    >
      <div
        class="mb-3"
      >
        <input
          class="form-control mr-3"
          placeholder="I am going..."
          required=""
          type="text"
          value=""
        />
      </div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        add
      </button>
    </form>
  </div>
  <div
    class="todo-active-tasks"
  >
    <div
      class="row"
    >
      <div
        class="col-1"
      >
        2
      </div>
      <div
        class="col"
      >
        <a
          class="todo-task"
          href="#"
        >
          second task
        </a>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col-1"
      >
        1
      </div>
      <div
        class="col"
      >
        <a
          class="todo-task"
          href="#"
        >
          first task
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`TodoBox 6`] = `
<div>
  <div
    class="mb-3"
  >
    <form
      class="todo-form"
    >
      <div
        class="mb-3"
      >
        <input
          class="form-control mr-3"
          placeholder="I am going..."
          required=""
          type="text"
          value=""
        />
      </div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        add
      </button>
    </form>
  </div>
  <div
    class="todo-active-tasks"
  >
    <div
      class="row"
    >
      <div
        class="col-1"
      >
        2
      </div>
      <div
        class="col"
      >
        <a
          class="todo-task"
          href="#"
        >
          second task
        </a>
      </div>
    </div>
  </div>
  <div
    class="todo-finished-tasks"
  >
    <div
      class="row"
    >
      <div
        class="col-1"
      >
        1
      </div>
      <div
        class="col"
      >
        <s>
          <a
            class="todo-task"
            href="#"
          >
            first task
          </a>
        </s>
      </div>
    </div>
  </div>
</div>
`;

exports[`TodoBox 7`] = `
<div>
  <div
    class="mb-3"
  >
    <form
      class="todo-form"
    >
      <div
        class="mb-3"
      >
        <input
          class="form-control mr-3"
          placeholder="I am going..."
          required=""
          type="text"
          value=""
        />
      </div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        add
      </button>
    </form>
  </div>
  <div
    class="todo-active-tasks"
  >
    <div
      class="row"
    >
      <div
        class="col-1"
      >
        2
      </div>
      <div
        class="col"
      >
        <a
          class="todo-task"
          href="#"
        >
          second task
        </a>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col-1"
      >
        1
      </div>
      <div
        class="col"
      >
        <a
          class="todo-task"
          href="#"
        >
          first task
        </a>
      </div>
    </div>
  </div>
</div>
`;
